File: src/improvement/dep_updates.py
"""Automated dependency updates with verification.

Scans for outdated dependencies, classifies by update type (major/minor/patch),
and provides update+test+rollback flow.
"""

from __future__ import annotations

import json
import os
import re
import subprocess
from dataclasses import dataclass
from enum import Enum


class UpdateType(Enum):
    PATCH = "patch"
    MINOR = "minor"
    MAJOR = "major"


@dataclass
class OutdatedDep:
    name: str
    current: str
    latest: str
    update_type: UpdateType


@dataclass
class UpdateResult:
    dep: OutdatedDep
    success: bool
    tests_passed: bool
    message: str


def classify_update(current: str, latest: str) -> UpdateType:
    """Classify version change as patch, minor, or major."""
    def parse(v: str) -> tuple[int, ...] | None:
        parts = re.findall(r'\d+', v)
        return tuple(int(p) for p in parts[:3]) if len(parts) >= 3 else None

    curr = parse(current)
    lat = parse(latest)

    if curr is None or lat is None:
        return UpdateType.MAJOR

    if curr[0] != lat[0]:
        return UpdateType.MAJOR
    if curr[1] != lat[1]:
        return UpdateType.MINOR
    return UpdateType.PATCH


def scan_python_outdated(project_dir: str) -> list[OutdatedDep]:  # pragma: no cover — subprocess
    """Scan for outdated Python packages using pip."""
    try:
        result = subprocess.run(
            ["pip", "list", "--outdated", "--format=json"],
            capture_output=True,
            timeout=60,
            cwd=project_dir,
        )
        if result.returncode != 0:
            return []

        packages = json.loads(result.stdout.decode())
        return [
            OutdatedDep(
                name=p["name"],
                current=p["version"],
                latest=p["latest_version"],
                update_type=classify_update(p["version"], p["latest_version"]),
            )
            for p in packages
        ]
    except (subprocess.TimeoutExpired, json.JSONDecodeError, FileNotFoundError):
        return []


def update_dependency(  # pragma: no cover — subprocess
    dep: OutdatedDep,
    project_dir: str,
    test_command: list[str] | None = None,
) -> UpdateResult:
    """Update a single dependency and verify with tests.

    Returns UpdateResult with success status.
    """
    # Install the update
    try:
        install = subprocess.run(
            ["pip", "install", f"{dep.name}=={dep.latest}"],
            capture_output=True,
            timeout=120,
            cwd=project_dir,
        )
        if install.returncode != 0:
            return UpdateResult(
                dep=dep, success=False, tests_passed=False,
                message=f"pip install failed: {install.stderr.decode()[:200]}",
            )
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return UpdateResult(
            dep=dep, success=False, tests_passed=False,
            message="pip install timed out or not found",
        )

    # Run tests if command provided
    if test_command:
        try:
            test_result = subprocess.run(
                test_command,
                capture_output=True,
                timeout=300,
                cwd=project_dir,
            )
            if test_result.returncode != 0:
                # Rollback
                subprocess.run(
                    ["pip", "install", f"{dep.name}=={dep.current}"],
                    capture_output=True,
                    timeout=120,
                    cwd=project_dir,
                )
                return UpdateResult(
                    dep=dep, success=False, tests_passed=False,
                    message=f"Tests failed after update, rolled back to {dep.current}",
                )
        except subprocess.TimeoutExpired:
            subprocess.run(
                ["pip", "install", f"{dep.name}=={dep.current}"],
                capture_output=True,
                timeout=120,
                cwd=project_dir,
            )
            return UpdateResult(
                dep=dep, success=False, tests_passed=False,
                message="Tests timed out, rolled back",
            )

    return UpdateResult(
        dep=dep, success=True, tests_passed=True,
        message=f"Updated {dep.name} from {dep.current} to {dep.latest}",
    )

File: rust/src/improvement/dep_updates.rs
//! Dependency update classification — major/minor/patch.

use regex::Regex;
use serde::{Deserialize, Serialize};

/// Type of version update.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum UpdateType {
    Patch,
    Minor,
    Major,
}

/// An outdated dependency.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OutdatedDep {
    pub name: String,
    pub current: String,
    pub latest: String,
    pub update_type: UpdateType,
}

/// Result of updating a dependency.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateResult {
    pub dep_name: String,
    pub success: bool,
    pub tests_passed: bool,
    pub message: String,
}

/// Parse a version string into (major, minor, patch).
fn parse_version(v: &str) -> Option<(u64, u64, u64)> {
    let re = Regex::new(r"\d+").unwrap();
    let parts: Vec<u64> = re
        .find_iter(v)
        .take(3)
        .filter_map(|m| m.as_str().parse().ok())
        .collect();
    if parts.len() >= 3 {
        Some((parts[0], parts[1], parts[2]))
    } else {
        None
    }
}

/// Classify version change as patch, minor, or major.
pub fn classify_update(current: &str, latest: &str) -> UpdateType {
    let curr = parse_version(current);
    let lat = parse_version(latest);

    match (curr, lat) {
        (Some((cm, _, _)), Some((lm, _, _))) if cm != lm => UpdateType::Major,
        (Some((_, cm, _)), Some((_, lm, _))) if cm != lm => UpdateType::Minor,
        (Some(_), Some(_)) => UpdateType::Patch,
        _ => UpdateType::Major,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_classify_patch() {
        assert_eq!(classify_update("1.2.3", "1.2.4"), UpdateType::Patch);
    }

    #[test]
    fn test_classify_minor() {
        assert_eq!(classify_update("1.2.3", "1.3.0"), UpdateType::Minor);
    }

    #[test]
    fn test_classify_major() {
        assert_eq!(classify_update("1.2.3", "2.0.0"), UpdateType::Major);
    }

    #[test]
    fn test_classify_unparseable() {
        assert_eq!(classify_update("abc", "def"), UpdateType::Major);
    }

    #[test]
    fn test_outdated_dep_serde() {
        let dep = OutdatedDep {
            name: "serde".into(),
            current: "1.0.0".into(),
            latest: "1.1.0".into(),
            update_type: UpdateType::Minor,
        };
        let json = serde_json::to_string(&dep).unwrap();
        let parsed: OutdatedDep = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed.name, "serde");
    }
}
